File: RazpoznavanjeMelodije/alvas/EffectsCs/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using Alvas.Audio;
using System.IO;
using System.Runtime.InteropServices;

namespace EffectsCs
{
    public partial class MainForm : Form
    {
        public MainForm()
        {
            InitializeComponent();
            Init();
        }

        private void Init()
        {
            ofdAudio.Filter = "*.wav|*.wav|*.mp3|*.mp3|*.avi|*.avi|*.wma;*.wmv;*.asf;*.mpg;*.aif;*.au;*.snd;*.mid;*.rmi;*.ogg;*.flac;*.cda;*.ac3;*.dts;*.mka;*.mkv;*.mpc;*.m4a;*.aac;*.mpa;*.mp2;*.m1a;*.m2a|*.wma;*.wmv;*.asf;*.mpg;*.aif;*.au;*.snd;*.mid;*.rmi;*.ogg;*.flac;*.cda;*.ac3;*.dts;*.mka;*.mkv;*.mpc;*.m4a;*.aac;*.mpa;*.mp2;*.m1a;*.m2a|*.*|*.*";
            btnPlay.Enabled = false;
            btnStop.Enabled = false;
        }

        PlayerEx plex = new PlayerEx();

        private void btnPlay_Click(object sender, EventArgs e)
        {
            AudioEffect af = null;
            foreach (Control ctrl in this.gbEffects.Controls)
            {
                RadioButton rb = ctrl as RadioButton;
                if (rb != null && rb.Checked)
                {
                    switch (rb.Text)
                    {
                        case "Chorus" :
                            af = AudioEffect.CreateChorusAudioEffect();
                            break;
                        case "Compressor":
                            af = AudioEffect.CreateCompressorAudioEffect();
                            break;
                        case "Distortion":
                            af = AudioEffect.CreateDistortionAudioEffect();
                            break;
                        case "Echo":
                            af = AudioEffect.CreateEchoAudioEffect();
                            break;
                        case "Flanger":
                            af = AudioEffect.CreateFlangerAudioEffect();
                            break;
                        case "Gargle":
                            af = AudioEffect.CreateGargleAudioEffect();
                            break;
                        case "I3DL2Reverb":
                            af = AudioEffect.CreateI3DL2ReverbAudioEffect();
                            break;
                        case "ParamEq":
                            af = AudioEffect.CreateParamEqAudioEffect();
                            break;
                        case "WavesReverb":
                            af = AudioEffect.CreateWavesReverbAudioEffect();
                            break;
                        default://None
                            break;
                    }
                    break;
                }
            }
            Play(arw, af);
            gbEffects.Enabled = false;
            btnPlay.Enabled = false;
            btnOpen.Enabled = false;
            btnStop.Enabled = true;
        }

        public void Play(IAudioReader wr, AudioEffect af)
        {
            //string fileName = @"c:\Wint\audio\NAudio-1-3\Source Code\MixDiff\bin\Debug\Windows XP Startup.wav";//1,16,8000.wav//G:\10\mp3\2,16,44100.wav
            //WaveReader wr = new WaveReader(File.OpenRead(fileName));
            IntPtr format = wr.ReadFormat();
            WaveFormat wf1 = AudioCompressionManager.GetWaveFormat(format);
            Console.WriteLine("{0},{1},{2}-{3}", wf1.nChannels, wf1.wBitsPerSample, wf1.nSamplesPerSec, wf1.wFormatTag);
            byte[] data = wr.ReadData();
            if (wf1.wFormatTag != 1)
            {
                IntPtr formatNew = IntPtr.Zero;
                byte[] dataNew = null;
                AudioCompressionManager.ToPcm(format, data, ref formatNew, ref dataNew);
                format = formatNew;
                data = dataNew;
                WaveFormat wf2 = AudioCompressionManager.GetWaveFormat(format);
                Console.WriteLine("{0},{1},{2}-{3}", wf2.nChannels, wf2.wBitsPerSample, wf2.nSamplesPerSec, wf2.wFormatTag);
            }
            else if (wf1.wBitsPerSample != 16)
            {
                WaveFormat wf = AudioCompressionManager.GetWaveFormat(format);
                IntPtr formatNew = AudioCompressionManager.GetPcmFormat(wf.nChannels, 16, wf.nSamplesPerSec);
                byte[] dataNew = AudioCompressionManager.Convert(format, formatNew, data, false);
                format = formatNew;
                data = dataNew;
                WaveFormat wf2 = AudioCompressionManager.GetWaveFormat(format);
                Console.WriteLine("{0},{1},{2}-{3}", wf2.nChannels, wf2.wBitsPerSample, wf2.nSamplesPerSec, wf2.wFormatTag);
            }
            //wr.Close();
            if (af != null)
            {
                bool hasProcessInPlace = af.HasProcessInPlace;
                //af.GetSupportedOutputFormats();
                GCHandle src = GCHandle.Alloc(data, GCHandleType.Pinned);
                IntPtr formatPtr = src.AddrOfPinnedObject();
                bool res = af.ProcessInPlace(format, data);
                src.Free();
                if (!res)
                {
                    MessageBox.Show("Unable to convert the audio data");
                    return;
                }
            }
            if (plex.State != DeviceState.Closed)
            {
                plex.ClosePlayer();
            }
            Console.WriteLine(plex.State);
            plex.OpenPlayer(format);
            plex.AddData(data);
            plex.StartPlay();
        }

        private void btnStop_Click(object sender, EventArgs e)
        {
            if (plex.State != DeviceState.Closed)
            {
                plex.ClosePlayer();
            }
            gbEffects.Enabled = true;
            btnPlay.Enabled = true;
            btnOpen.Enabled = true;
            btnStop.Enabled = false;
            Console.WriteLine("{0} STOP", plex.State);
        }

        IAudioReader arw;

        private void btnOpen_Click(object sender, EventArgs e)
        {
            if (ofdAudio.ShowDialog() == DialogResult.OK)
            {
                if (arw != null)
                {
                    arw.Close();
                    arw = null;
                }
                string fileName = ofdAudio.FileName;
                arw = null;
                switch (Path.GetExtension(fileName.ToLower()))
                {
                    case ".avi":
                        arw = new AviReader(File.Open(fileName, FileMode.Open, FileAccess.ReadWrite));
                        if (!((AviReader)arw).HasAudio)
                        {
                            MessageBox.Show(string.Format("'{0}' file is not contains audio data", fileName));
                            return;
                        }
                        break;
                    case ".au":
                    case ".snd":
                        arw = new AuReader(File.OpenRead(fileName));
                        break;
                    case ".wav":
                        arw = new WaveReadWriter(File.Open(fileName, FileMode.Open, FileAccess.ReadWrite));
                        break;
                    case ".mp3":
                        arw = new Mp3ReadWriter(File.Open(fileName, FileMode.Open, FileAccess.ReadWrite));
                        break;
                    default:
                        arw = new DsReader(fileName);
                        if (!((DsReader)arw).HasAudio)
                        {
                            arw = null;
                            MessageBox.Show(string.Format("'{0}' file is not contains audio data", fileName));
                        }
                        break;
                }
                btnPlay.Enabled = arw != null;
            }
        }

    }
}

File: RazpoznavanjeMelodije/alvas/EffectsVb/MainForm.vb
Imports Alvas.Audio
Imports System.IO
Imports System.Runtime.InteropServices

Public Class MainForm

    'Private Sub btnOpen_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnOpen.Click

    'End Sub

    'Private Sub btnPlay_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnPlay.Click

    'End Sub

    'Private Sub btnStop_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnStop.Click

    'End Sub

    Public Sub New()
        InitializeComponent()
        Init()
    End Sub

    Private Sub Init()
        ofdAudio.Filter = "*.wav|*.wav|*.mp3|*.mp3|*.avi|*.avi|*.wma;*.wmv;*.asf;*.mpg;*.aif;*.au;*.snd;*.mid;*.rmi;*.ogg;*.flac;*.cda;*.ac3;*.dts;*.mka;*.mkv;*.mpc;*.m4a;*.aac;*.mpa;*.mp2;*.m1a;*.m2a|*.wma;*.wmv;*.asf;*.mpg;*.aif;*.au;*.snd;*.mid;*.rmi;*.ogg;*.flac;*.cda;*.ac3;*.dts;*.mka;*.mkv;*.mpc;*.m4a;*.aac;*.mpa;*.mp2;*.m1a;*.m2a|*.*|*.*"
        btnPlay.Enabled = False
        btnStop.Enabled = False
    End Sub

    Private plex As New PlayerEx()

    Private Sub btnPlay_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnPlay.Click
        Dim af As AudioEffect = Nothing
        For Each ctrl As Control In Me.gbEffects.Controls
            Dim rb As RadioButton = TryCast(ctrl, RadioButton)
            If rb IsNot Nothing AndAlso rb.Checked Then
                Select Case rb.Text
                    Case "Chorus"
                        af = AudioEffect.CreateChorusAudioEffect()
                        Exit Select
                    Case "Compressor"
                        af = AudioEffect.CreateCompressorAudioEffect()
                        Exit Select
                    Case "Distortion"
                        af = AudioEffect.CreateDistortionAudioEffect()
                        Exit Select
                    Case "Echo"
                        af = AudioEffect.CreateEchoAudioEffect()
                        Exit Select
                    Case "Flanger"
                        af = AudioEffect.CreateFlangerAudioEffect()
                        Exit Select
                    Case "Gargle"
                        af = AudioEffect.CreateGargleAudioEffect()
                        Exit Select
                    Case "I3DL2Reverb"
                        af = AudioEffect.CreateI3DL2ReverbAudioEffect()
                        Exit Select
                    Case "ParamEq"
                        af = AudioEffect.CreateParamEqAudioEffect()
                        Exit Select
                    Case "WavesReverb"
                        af = AudioEffect.CreateWavesReverbAudioEffect()
                        Exit Select
                    Case Else
                        'None
                        Exit Select
                End Select
                Exit For
            End If
        Next
        Play(arw, af)
        gbEffects.Enabled = False
        btnPlay.Enabled = False
        btnOpen.Enabled = False
        btnStop.Enabled = True
    End Sub

    Public Sub Play(ByVal wr As IAudioReader, ByVal af As AudioEffect)
        'string fileName = @"c:\Wint\audio\NAudio-1-3\Source Code\MixDiff\bin\Debug\Windows XP Startup.wav";//1,16,8000.wav//G:\10\mp3\2,16,44100.wav
        'WaveReader wr = new WaveReader(File.OpenRead(fileName));
        Dim format As IntPtr = wr.ReadFormat()
        Dim wf1 As WaveFormat = AudioCompressionManager.GetWaveFormat(format)
        Console.WriteLine("{0},{1},{2}-{3}", wf1.nChannels, wf1.wBitsPerSample, wf1.nSamplesPerSec, wf1.wFormatTag)
        Dim data As Byte() = wr.ReadData()
        If wf1.wFormatTag <> 1 Then
            Dim formatNew As IntPtr = IntPtr.Zero
            Dim dataNew As Byte() = Nothing
            AudioCompressionManager.ToPcm(format, data, formatNew, dataNew)
            format = formatNew
            data = dataNew
            Dim wf2 As WaveFormat = AudioCompressionManager.GetWaveFormat(format)
            Console.WriteLine("{0},{1},{2}-{3}", wf2.nChannels, wf2.wBitsPerSample, wf2.nSamplesPerSec, wf2.wFormatTag)
        ElseIf wf1.wBitsPerSample <> 16 Then
            Dim wf As WaveFormat = AudioCompressionManager.GetWaveFormat(format)
            Dim formatNew As IntPtr = AudioCompressionManager.GetPcmFormat(wf.nChannels, 16, wf.nSamplesPerSec)
            Dim dataNew As Byte() = AudioCompressionManager.Convert(format, formatNew, data, False)
            format = formatNew
            data = dataNew
            Dim wf2 As WaveFormat = AudioCompressionManager.GetWaveFormat(format)
            Console.WriteLine("{0},{1},{2}-{3}", wf2.nChannels, wf2.wBitsPerSample, wf2.nSamplesPerSec, wf2.wFormatTag)
        End If
        'wr.Close();
        If af IsNot Nothing Then
            Dim hasProcessInPlace As Boolean = af.HasProcessInPlace
            'af.GetSupportedOutputFormats();
            Dim src As GCHandle = GCHandle.Alloc(data, GCHandleType.Pinned)
            Dim formatPtr As IntPtr = src.AddrOfPinnedObject()
            Dim res As Boolean = af.ProcessInPlace(format, data)
            src.Free()
            If Not res Then
                MessageBox.Show("Unable to convert the audio data")
                Return
            End If
        End If
        If plex.State <> DeviceState.Closed Then
            plex.ClosePlayer()
        End If
        Console.WriteLine(plex.State)
        plex.OpenPlayer(format)
        plex.AddData(data)
        plex.StartPlay()
    End Sub

    Private Sub btnStop_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnStop.Click
        If plex.State <> DeviceState.Closed Then
            plex.ClosePlayer()
        End If
        gbEffects.Enabled = True
        btnPlay.Enabled = True
        btnOpen.Enabled = True
        btnStop.Enabled = False
        Console.WriteLine("{0} STOP", plex.State)
    End Sub

    Private arw As IAudioReader

    Private Sub btnOpen_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnOpen.Click
        If ofdAudio.ShowDialog() = DialogResult.OK Then
            If arw IsNot Nothing Then
                arw.Close()
                arw = Nothing
            End If
            Dim fileName As String = ofdAudio.FileName
            arw = Nothing
            Select Case Path.GetExtension(fileName.ToLower())
                Case ".au"
                Case ".snd"
                    arw = New AuReader(File.OpenRead(fileName))
                    Exit Select
                Case ".avi"
                    arw = New AviReader(File.Open(fileName, FileMode.Open, FileAccess.ReadWrite))
                    If Not DirectCast(arw, AviReader).HasAudio Then
                        MessageBox.Show(String.Format("'{0}' file is not contains audio data", fileName))
                        Return
                    End If
                    Exit Select
                Case ".wav"
                    arw = New WaveReadWriter(File.Open(fileName, FileMode.Open, FileAccess.ReadWrite))
                    Exit Select
                Case ".mp3"
                    arw = New Mp3ReadWriter(File.Open(fileName, FileMode.Open, FileAccess.ReadWrite))
                    Exit Select
                Case Else
                    arw = New DsReader(fileName)
                    If Not DirectCast(arw, DsReader).HasAudio Then
                        arw = Nothing
                        MessageBox.Show(String.Format("'{0}' file is not contains audio data", fileName))
                    End If
                    Exit Select
            End Select
            btnPlay.Enabled = arw IsNot Nothing
        End If
    End Sub

End Class
